### PR TITLE
En 8340/selection at end of epoch

### PR DESCRIPTION
**Pull request by sasurobert**
logic has been implemented - tests and fixes of broken tests in next commits

### CODE REVIEW

File: epochStart/metachain/stakingDataProvider.go [hunk 1]
```diff
@@ -239,32 +264,119 @@ func (sdp *stakingDataProvider) getTopUpValueAndTotalStaked(validatorAddress str
...
+
+	if int64(len(selectedKeys)) >= numToSelect {
+		return selectedKeys[:numToSelect]
+	}
+
+	eligibleKeys := sortedKeys[string(core.EligibleList)]
+	if len(waitingKeys) > 0 {
+		selectedKeys = append(selectedKeys, eligibleKeys...)
+	}
+
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/stakingDataProvider.go:359]: maybe yes.

File: epochStart/metachain/stakingDataProvider.go [hunk 2]
```diff
@@ -239,32 +264,119 @@ func (sdp *stakingDataProvider) getTopUpValueAndTotalStaked(validatorAddress str
...
+func (sdp *stakingDataProvider) ComputeUnQualifiedNodes(validatorInfos map[uint32][]*state.ValidatorInfo) ([][]byte, [][]byte, error) {
+	sdp.mutStakingData.Lock()
+	defer sdp.mutStakingData.Unlock()
+
+	ownersWithNotEnoughFunds := make([][]byte, 0)
+	keysToUnStake := make([][]byte, 0)
+	mapBLSKeyStatus := createMapBLSKeyStatus(validatorInfos)
+	for ownerAddress, stakingInfo := range sdp.cache {
+		numRegisteredKeys := int64(len(stakingInfo.blsKeys))
+		maxQualified := big.NewInt(0).Div(stakingInfo.totalStaked, sdp.minNodePrice)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/stakingDataProvider.go:305]: done

File: epochStart/metachain/systemSCs.go [hunk 3]
```diff
@@ -190,16 +207,162 @@ func (s *systemSCProcessor) ProcessSystemSmartContract(validatorInfos map[uint32
...
+	validatorInfos map[uint32][]*state.ValidatorInfo,
+	epoch uint32,
+) (uint32, error) {
+	nodesToUnStake, _, err := s.stakingDataProvider.ComputeUnQualifiedNodes(validatorInfos)
+	if err != nil {
+		return 0, err
+	}
+
+	var vmOutput *vmcommon.VMOutput
+	for _, blsKey := range nodesToUnStake {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:252]: done

File: epochStart/metachain/stakingDataProvider.go [hunk 4]
```diff
@@ -239,32 +269,123 @@ func (sdp *stakingDataProvider) getTopUpValueAndTotalStaked(validatorAddress str
...
 	}
 
-	totalStakedValue, ok := big.NewInt(0).SetString(string(topUpBytes[1]), conversionBase)
+	totalStakedValue, ok := big.NewInt(0).SetString(string(vmOutput.ReturnData[1]), conversionBase)
 	if !ok {
-		return nil, nil, fmt.Errorf("%w, error: totalStaked string returned is not a number", epochStart.ErrExecutingSystemScCode)
+		return nil, nil, nil, fmt.Errorf("%w, error: totalStaked string returned is not a number", epochStart.ErrExecutingSystemScCode)
+	}
+
+	return topUpValue, totalStakedValue, vmOutput.ReturnData[2:], nil
```
Comment (AdoAdoAdo) [inline epochStart/metachain/stakingDataProvider.go:297]: Are the "leaving" keys also among the returned keys here?

File: epochStart/metachain/systemSCs.go [hunk 5]
```diff
@@ -910,7 +1101,8 @@ func (s *systemSCProcessor) EpochConfirmed(epoch uint32) {
 	s.flagDelegationEnabled.Toggle(epoch == s.delegationEnableEpoch)
 	log.Debug("systemSCProcessor: delegation", "enabled", epoch >= s.delegationEnableEpoch)
 
-	s.flagStakingV2Enabled.Toggle(epoch == s.stakingV2EnableEpoch)
+	s.flagSetOwnerEnabled.Toggle(epoch == s.stakingV2EnableEpoch)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:1104]: One has to be called only once. The other at every start of epoch. Activated by the same enabled epoch value

File: epochStart/metachain/systemSCs.go [hunk 6]
```diff
@@ -876,6 +1055,18 @@ func (s *systemSCProcessor) addNewlyStakedNodesToValidatorTrie(returnData [][]by
 		if err != nil {
 			return err
 		}
+
+		validatorInfo := &state.ValidatorInfo{
+			PublicKey:       blsKey,
+			ShardId:         peerAcc.GetShardId(),
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:1061]: Default is 0. It does not really matter :) as shuffler will put the actual ID a little later

File: epochStart/metachain/stakingDataProvider.go (hunk 4 above)
Comment (sasurobert) [reply epochStart/metachain/stakingDataProvider.go:297]: All the blsKeys are returned. Everything which was registered

File: epochStart/metachain/systemSCs.go [hunk 7]
```diff
@@ -809,7 +980,11 @@ func (s *systemSCProcessor) updateSystemSCContractsCode(contractMetadata []byte)
 	return nil
 }
 
-func (s *systemSCProcessor) stakeNodesFromWaitingList(nodesToStake uint32, nonce uint64) error {
+func (s *systemSCProcessor) stakeNodesFromWaitingList(
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:983]: Yes. Will do

[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go]: done

**Review by iulianpascalau — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Pull request merged**